scripts/install-nix-from-closure.sh

```diff
@@ -1,149 +1,602 @@
+{
+print() {
+    # Unified output function
+    # Using printf, because it is more portable across all platforms than echo, and it is important for Nix install
+    #
+    if [ -z "$message" ]; then
+        message="$1"
+    fi
+    if [ -z "$color" ]; then
+        color="$reset"
+    fi
+    if [ -z "$prefix" ]; then
+        prefix='Info'
+    fi
+    # This line makes all prints in script
+    # Form:
+    ### Application: Prefix: Body of message
+    printf '%s%s: %s: %s%s\n' "$color" "$appname" "$prefix" "$message" "$reset"
+
+    # Output is done, unset variables that could've been set by parent functions
+    unset message
+    unset color
+    unset prefix
+}
+
+notice() {
+    message="$1"
+    color="$green"
+    prefix='Notice'
+    print
+}
+
+warning() {
+    message="$1"
+    color="$yellow"
+    prefix='Warning'
+    >&2 print
+}
+
+error() {
+    message="$1"
+    exitSign="$2"
+    color="$red"
+    prefix='Error'
+    >&2 print
+    if [ -z "$exitSign" ]; then
+        exit 1
+    fi
+}
+
+errorRevert() {
+    message="$1"
+    exitSign="$2"
+    color="$red"
+    prefix='Error'
+    >&2 print
+}
+}
+###############################
+###  Program stage functions
+###############################
+{
+checkingRequirements() {    # This function must have checks only. Without making ANY changes to the system.
+    if ! [ -e "$self/.reginfo" ]; then
+        error "
+
+    Installer is incomplete ($self/.reginfo is missing)
+    "
+    fi
 
-set -e
+    if [ "$(id -u)" -eq 0 ]; then
+        notice "
 
-dest="/nix"
-self="$(dirname "$0")"
-nix="@nix@"
-cacert="@cacert@"
+    Install executed for ROOT.
 
+    Doing classic Linux package manager mode.
+    In Nix this mode is called: single-user mode for root.
 
-if ! [ -e "$self/.reginfo" ]; then
-    echo "$0: incomplete installer (.reginfo is missing)" >&2
-fi
+    Nix can do multi-user mode, and manage package trees for users independently.
+    "
+    fi
 
-if [ -z "$USER" ]; then
-    echo "$0: \$USER is not set" >&2
-    exit 1
-fi
+    # In case if USER not set
+    if [ -z "$USER" ]; then
+        notice "
 
-if [ -z "$HOME" ]; then
-    echo "$0: \$HOME is not set" >&2
-    exit 1
-fi
+    Environment variable USER is not set.
+    Maybe running inside container.
+    "
+        USER="$(whoami)"    # Taking username from UID
+        print "Detected username: $USER"
+    fi
 
-# macOS support for 10.10 or higher
-if [ "$(uname -s)" = "Darwin" ]; then
-    if [ $(($(sw_vers -productVersion | cut -d '.' -f 2))) -lt 10 ]; then
-        echo "$0: macOS $(sw_vers -productVersion) is not supported, upgrade to 10.10 or higher"
-        exit 1
+    if [ -z "$HOME" ]; then
+        error "
+
+    Environment variable HOME is not set.
+    "
     fi
 
-    printf '\e[1;31mSwitching to the Multi-User Darwin Installer\e[0m\n'
-    exec "$self/install-darwin-multi-user"
-    exit 0
-fi
+    # macOS support for 10.10 or higher
+    if [[ "$(uname -s)" = "Darwin" ]]; then
```

Comment:
This is @grahamc code.
`[[` is a Bash function, and I think on macOS default shell is Bash.

`sh` just skips blocks it does not understand. And since it is macOS bridge block, Linux functionality is not touched.
Shell just skips it, as it was doing with this same code in versions before.

Reply:
It is not my code.

I changed it, and than changed back as it was.
And since it is macOS bridge, macOS uses bash.

Yes, ok, will be
```shell
if [ "$(uname -s)" = "Darwin" ]; then
```

Reply:
Done. In: https://github.com/NixOS/nix/pull/1565/commits/bfec07693da39f2cc618563794657f2ca751a9ae

But `[[ ]]` under that is not fixed. It is macOS related code, I can not try it to understand why is it done so.
```shell
if [[ $(($(sw_vers -productVersion | cut -d '.' -f 2))) -lt 10 ]]; then
```
Is done to get a release number (aka 10.12.02), and get second number. And convert is from string, maybe with possible spaces, to be evaluated as number and `<` compared.

It also can be:
```shell
readonly macosVersion=$(sw_vers -productVersion)
if [ $(printf "$macosVersion" | cut -d '.' -f 1) = 10 ]; then # since all mac versions are 10.x, and now are 10.13.. they can switch to next number in the future
    if [ $(printf "$macosVersion" | cut -d '.' -f 2) -lt 10 ]; then
    ...
    fi
fi
```
But maybe I miss something in macOS.
I really don't want to change code for the system I don't know much about.